DocGen4/Output/References.lean

```diff
@@ -0,0 +1,77 @@
+import DocGen4.Output.Template
+
+/-!
+
+# Generic functions for references support
+
+This file contains functions for references support,
+independent of actual implementation.
+
+The main function is `DocGen4.preprocessBibFile` which preprocess
+the contents of bib file using user provided `process` function,
+and save the bib file and processed json file to output directory.
+
+Currently we have an implementation of `process` function using `pybtex`,
+which is `DocGen4.Pybtex.process`. In the future this may be replaced by
+a pure Lean implementation.
+
+-/
+
+open Lean DocGen4 Output
+
+namespace DocGen4
+
+/-- Preprocess (using the user provided `process` function)
+and save the bib file to the output path. -/
+def preprocessBibFile (contents : String) (process : String → IO (Array BibItem)) : IO Unit := do
+  -- create directories
+  IO.FS.createDirAll basePath
+  IO.FS.createDirAll declarationsBasePath
+  -- erase all files
+  IO.FS.writeFile (basePath / "references.bib") contents
+  IO.FS.writeFile (declarationsBasePath / "references.json") "[]"
+  -- if contents is empty, just do nothing
+  if contents.trim.isEmpty then
+    return
+  -- run the user provided process function
+  let items ← process contents
+  -- save the result
+  IO.FS.writeFile (declarationsBasePath / "references.json") (toString (toJson items))
+
+namespace Output
+
+open scoped DocGen4.Jsx
+
+def refItem (ref : BibItem) (backrefs : Array BackrefItem) : BaseHtmlM Html := do
+  let backrefs := backrefs.filter (fun x => x.citekey == ref.citekey)
+  let toHtml (i : Nat) : BaseHtmlM (Array Html) := do
+    let .some backref := backrefs.get? i | unreachable!
```

Comment:
Use `backrefs[i]!`

Reply:
This is not working (same as `get!`), mostly because there is no `Inhabited BackrefItem`.

Reply:
Then you should just be able to derive it and it works no? I don't see a component of that structure that should block this.